MaxText/maxtext_utils.py

```
@@ -372,19 +401,152 @@ def assert_params_sufficiently_sharded(params, mesh, tolerance):
+    if not is_sharded_on_all_target_axis:
+      unsharded_params_total_size += p_leaf.size  # Add to total unsharded parameter size
+      unsharded_axes = set(valid_target_mesh_axes) - set(mesh_axes_used)
+      # Add detailed info to list of problematic tensors
+      problematic_tensors_details.append(
+          {
+              "name": param_name_str,  # Tensor name
+              "size": p_leaf.size,  # tensor size
+              "spec": str(current_sharding_spec),  # Tensor sharding spec as string
+              "available_axes": sorted(list(valid_target_mesh_axes)),  # Axes that could be used for sharding
+              "unsharded_axes": sorted(list(unsharded_axes)),  # Unsharded axes
+          }
+      )
+  # Return the total size of unsharded parameters and the list of problematic tensors.
+  return unsharded_params_total_size, problematic_tensors_details  # Return results
+
+
+def _raise_if_unsharded_exceeds_tolerance(unsharded_size, total_size, tolerance, problematic_tensors_details):
+  """
+  Raises an AssertionError if the percentage of unsharded parameters exceeds the given tolerance.
+
+  This function calculates the proportion of model parameters that are unsharded
+  and compares it against a specified tolerance. If the tolerance is exceeded,
+  it constructs and raises a detailed error message.
+
+  Args:
+    unsharded_size: The total size of parameters not sharded on target axes.
+    total_size: The total size of all parameters in the model.
+    tolerance: A float (e.g., 0.05 for 5%) representing the maximum allowed percentage of unsharded parameters.
+    problematic_tensors_details: A list of details about the unsharded tensors,
+    used to generate an informative error message.
+
+  Raises:
+    AssertionError: If the percentage of unsharded parameters is greater than the tolerance.
+  """
+  # Calculate the percentage of unsharded parameters.
+  unsharded_param_perc = unsharded_size / total_size if total_size > 0 else 0.0
```

Should we raise an error if total_size is <=0? instead of setting unsharded_param_perc=0?

---

The current code treats a model with no parameters as a trivial case where there are no sharding issues to report. It avoids a "division by zero" error and correctly determines the percentage of unsharded parameters is 0.

The total_size is calculated from max_utils.calculate_bytes_from_pytree(params). This variable represents the total size of all parameters in the model. If total_size is 0, it means the model has no parameters. While this might seem unusual for a typical large model, it is a perfectly valid scenario in several common situations such as unit tests. We might have unit tests that check the behavior of the training loop or other components with an empty model to isolate the logic being tested. In some possible valid cases, a model with no parameters has, by definition, zero unsharded parameters.

If we try to raise an error when total_size is 0, it might cause the program to crash in some legitimate scenarios (e.g., unit tests). This would make the assertion utility less robust. Since this assertion is aiming to check for inefficient sharding, not sure whether it is necessary to check the empty model parameter.

---

> While this might seem unusual for a typical large model

Yeah, just a sanity check. Instead of silently pass `unsharded_param_perc = unsharded_size / total_size if total_size > 0 else 0.0`.

We could do:

```
if total_size <= 0:
  raise value error....

unsharded_param_perc = unsharded_size / total_size
```

> If we try to raise an error when total_size is 0, it might cause the program to crash in some legitimate scenarios (e.g., unit tests).

Do we run a model with total_size=0 in unit test?

---

No, actually we do not run a model with total_size = 0 in the unit test. 

Good suggestion! I will add this error assertion part instead of silently passing.